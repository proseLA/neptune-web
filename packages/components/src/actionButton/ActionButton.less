.np-action-btn {
  display: inline-flex;
  height: var(--size-32);
  padding: 7px var(--size-12) var(--size-8);
  font-size: var(--font-size-14);
  line-height: var(--line-height-control);
  font-weight: var(--font-weight-semi-bold);
  color: var(--color-content-accent);
  background-color: var(--color-background-neutral);
  border: 1px solid transparent;
  border-radius: var(--radius-small);
  transition: all 0.15s ease-in-out;
  transition-property: color, border-color, background-color, box-shadow;

  &:not(:disabled) {
    &:hover {
      background-color: var(--color-background-accent);
    }

    &:focus,
    &:active {
      outline: none;
      border-color: var(--color-interactive-accent);
    }

    &:focus {
      box-shadow: 0 0 0 4px var(--color-background-accent);
    }
  }
}
