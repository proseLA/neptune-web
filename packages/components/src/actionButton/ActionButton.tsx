import classNames from 'classnames';
import { ButtonHTMLAttributes, MouseEvent, forwardRef } from 'react';

type ActionButtonProps = {
  className?: string;
  disabled?: boolean;
  text: string;
  onClick: (event: MouseEvent<HTMLButtonElement>) => void;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const ActionButton = forwardRef<HTMLButtonElement, ActionButtonProps>(
  (
    { disabled = false, className = undefined, onClick, text, ...rest }: ActionButtonProps,
    reference,
  ) => (
    <button
      ref={reference}
      type="button"
      className={classNames(`np-action-btn`, className)}
      disabled={disabled}
      onClick={onClick}
      {...rest}
    >
      {text}
    </button>
  ),
);

export default ActionButton;
