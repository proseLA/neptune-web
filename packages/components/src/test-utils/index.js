import { render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';

import { Provider } from '..';
import { DEFAULT_LOCALE } from '../common/locale';
import en from '../i18n/en.json';

/**
 * Custom `render` function which wraps passed elements in Provider component
 * Source: https://testing-library.com/docs/example-react-intl/#creating-a-custom-render-function
 *
 */
function customRender(
  ui,
  { locale = DEFAULT_LOCALE, messages = en, direction, ...renderOptions } = {},
) {
  // eslint-disable-next-line react/prop-types
  var Wrapper = ({ children }) => {
    return (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

/**
 * Custom `renderHook` function which wraps passed elements in Provider component
 * For more info: https://react-hooks-testing-library.com/usage/advanced-hooks#context
 *
 */
function customRenderHook(callback, { locale = DEFAULT_LOCALE, messages = en, direction } = {}) {
  return renderHook(callback, {
    // eslint-disable-next-line react/prop-types
    wrapper: ({ children }) => (
      <Provider i18n={{ locale, messages }} direction={direction}>
        {children}
      </Provider>
    ),
  });
}

export * from '@testing-library/react';
export { userEvent };
export { customRender as render, customRenderHook as renderHook };
