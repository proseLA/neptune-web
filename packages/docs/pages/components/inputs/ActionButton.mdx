import { LiveEditorBlock, GeneratePropsTable } from '../../../utils';
import { DocumentLink } from '../../../utils/pageUtils';
import { ActionButton, Link } from '@transferwise/components';
import code from '../../../liveEditorCode/actionButton.code';

Use action buttons for non-primary standalone actions or when there are multiple actions a user can choose from.

For primary actions, use a normal <DocumentLink href="components/inputs/Button">Button</DocumentLink>. Refer to the <Link href="https://neptune.wise.com/components/action-button/" target="_blank">action button usage guidelines</Link> for more information on the differences between the two components.

<LiveEditorBlock code={code} scope={{ ActionButton }} />
<GeneratePropsTable componentName="ActionButton" />

export const meta = {
  name: 'ActionButton',
  badge: {
    expiryDate: '2022-03-04',
    type: 'new',
  },
};
