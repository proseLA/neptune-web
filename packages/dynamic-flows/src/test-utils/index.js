import { Provider, translations as componentTranslations } from '@transferwise/components';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';

import { FetcherProviderFromBaseUrl } from '../common/contexts/fetcherContext';

const getMockFetchPromise = (status, jsonFn, delay, signal) => {
  const response = {
    status,
    json: jsonFn,
  };
  return new Promise((resolve) => {
    let aborted = false;
    if (signal) {
      signal.addEventListener('abort', () => {
        aborted = true;
      });
    }
    setTimeout(() => {
      return !aborted && resolve(response);
    }, delay);
  });
};

const wait = (t) => {
  return act(() => {
    return new Promise((resolve) => setTimeout(resolve, t));
  });
};

function mountWithProviders(
  component,
  locale = 'en-GB',
  // eslint-disable-next-line unicorn/no-object-as-default-parameter
  messages = { ...componentTranslations['en'] },
  baseUrl = 'https://test-url',
) {
  return mount(
    <Provider i18n={{ locale, messages }}>
      <FetcherProviderFromBaseUrl baseUrl={baseUrl}>{component}</FetcherProviderFromBaseUrl>
    </Provider>,
  );
}

function getI18n(
  locale = 'en-GB',
  // eslint-disable-next-line unicorn/no-object-as-default-parameter
  messages = { ...componentTranslations['en'] },
) {
  return { locale, messages };
}

export { mountWithProviders as mount, getMockFetchPromise, wait, getI18n };
