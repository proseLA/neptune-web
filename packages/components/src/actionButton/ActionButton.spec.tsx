import '@testing-library/jest-dom';
import { render, screen, userEvent } from '../test-utils';

import ActionButton from '.';

describe('ActionButton', () => {
  const props = {
    text: 'Personal profile',
    onClick: jest.fn(),
  };

  describe('by default', () => {
    beforeEach(() => {
      render(<ActionButton {...props} />);
    });

    it('renders the text', () => {
      expect(screen.getByText('Personal profile')).toBeInTheDocument();
    });

    it('is not disabled', () => {
      expect(screen.getByRole('button')).toBeEnabled();
    });

    it('renders the button', () => {
      expect(render(<ActionButton {...props} />).container).toMatchSnapshot();
    });
  });

  describe('button attributes', () => {
    it('disables the button if disabled', () => {
      render(<ActionButton {...props} disabled />);
      expect(screen.getByRole('button')).toBeDisabled();
    });

    it('passes through custom classes if set', () => {
      render(<ActionButton {...props} className="catsarethebest" />);
      expect(screen.getByRole('button')).toHaveClass('catsarethebest');
    });
  });

  describe('onClick', () => {
    it('calls onClick when clicked', () => {
      const onClick = jest.fn();
      render(<ActionButton {...props} onClick={onClick} />);
      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it('does not call onClick when clicked if disabled', () => {
      const onClick = jest.fn();
      render(<ActionButton {...props} disabled onClick={onClick} />);
      userEvent.click(screen.getByRole('button'));
      expect(onClick).toHaveBeenCalledTimes(0);
    });
  });
});
