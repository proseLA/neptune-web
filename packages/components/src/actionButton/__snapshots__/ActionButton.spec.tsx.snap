// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton by default renders the button 1`] = `
<div>
  <button
    class="np-action-btn"
    type="button"
  >
    Personal profile
  </button>
</div>
`;
